File: ios/Methods/MethodHandler.swift
//
// Copyright © 2023 Nevis Security AG. All rights reserved.
//

import NevisMobileAuthentication

protocol MethodHandler {
	func execute(using client: MobileAuthenticationClient, with message: ChannelInMessage) throws
}

extension MethodHandler {
	func validate<T>(message: ChannelInMessage) throws -> T {
		guard type(of: message) is T.Type else {
			throw PluginError.illegalArgument("Wrong type of message received!")
		}

		return message as! T
	}
}

File: android/src/main/java/ch/nevis/mobile/authentication/sdk/react/methods/MethodHandler.kt
/*
 * Copyright © 2023 Nevis Security AG. All rights reserved.
 */

package ch.nevis.mobile.authentication.sdk.react.methods

import ch.nevis.mobile.authentication.sdk.react.model.message.incoming.ChannelInMessage
import ch.nevis.mobile.sdk.api.MobileAuthenticationClient

abstract class MethodHandler {
    abstract fun execute(client: MobileAuthenticationClient, channelInMessage: ChannelInMessage)

    protected inline fun <reified T : ChannelInMessage> validate(message: ChannelInMessage): T {
        if (message is T) {
            return message
        }
        throw IllegalArgumentException("Invalid message parameter.")
    }
}

//fun MethodChannel.Result.deleteAuthenticatorErrorMessage(
//    deleteAuthenticatorError: DeleteAuthenticatorError,
//    exception: Exception
//) {
//    error(
//        PlatformChannelError.DELETE_AUTHENTICATOR.errorName,
//        exception.localizedMessage,
//        mapOf(
//            ChannelErrorKey.TYPE.keyName to deleteAuthenticatorError.type,
//            ChannelErrorKey.DATA.keyName to mapOf(
//                ChannelErrorKey.DESCRIPTION.keyName to exception.localizedMessage
//            )
//        )
//    )
//}
